In [22]:
import numpy as np 
import pandas as pd
import statsmodels.formula.api as smf
import seaborn as sns
import io
import scipy.stats as stats

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/10boston.csv')
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

## Спецификация
$$
RM= \beta_0 + \beta_1*CRIM + \beta_2*DIS
$$

In [5]:
price_eq = smf.ols(formula='RM~CRIM+DIS', data=df).fit()
price_eq.params

Intercept    6.152858
CRIM        -0.013488
DIS          0.047566
dtype: float64

In [6]:
# TSS
price_eq.centered_tss.round(3)

249.304

In [7]:
# ESS
price_eq.ess.round(3)

16.32

In [8]:
# RSS
RSS=price_eq.ssr.round(3)
RSS

232.984

$$
TSS=249.304, ESS=16.32, RSS=232.984
$$

### Задание 4

In [9]:
# R^2
price_eq.rsquared.round(3)

0.065

$$
R^2 = 0.065
$$

In [10]:
n=506
k=2
s=np.sqrt(RSS/(n-k-1))
round(s,3)

0.681

$$
s=0.681
$$

### Задание 5

In [11]:
# adj R^2
price_eq.rsquared_adj.round(3)

0.062

$$
R^2_{adj}= 0.062
$$

### Задание 6

## Интерпритация
- 1 При увеличении уровня преступности на душу населения на одну единицу, среднее количество комнат в квартире уменьшается на 0.013488 м^2, при прочих равных
- 2 При увеличении расстояния до пяти бостонских центров занятости на одну единицу, среднее количество комнат в квартире увеличивается на 0.047566 м^2, при прочих равных

### Задание 7

In [12]:
models = smf.ols(data=df, formula='RM~CRIM+DIS').fit()
models.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     RM   R-squared:                       0.065
Model:                            OLS   Adj. R-squared:                  0.062
Method:                 Least Squares   F-statistic:                     17.62
Date:                Fri, 24 Mar 2023   Prob (F-statistic):           4.03e-08
Time:                        19:48:40   Log-Likelihood:                -521.76
No. Observations:                 506   AIC:                             1050.
Df Residuals:                     503   BIC:                             1062.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.1529      0.072     85.311      0.000       6.011       6.295
CRIM          -0.0135      0.004     -3.544      0.000      -0.021      -0.006
DIS            0.0476      0.016      3.060      0.002       0.017       0.078
==============================================================================
Omnibus:                       55.829   Durbin-Watson:                   1.082
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              107.162
Skew:                           0.655   Prob(JB):                     5.37e-24
Kurtosis:                       4.835   Cond. No.                         22.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [14]:
model = smf.ols(data = df, formula = 'RM~CRIM+DIS').fit()
model.summary(alpha = 0.05).tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

In [18]:
data_string='''coef std.error
Intercept     6.1529  0.072
CRIM     -0.0135  0.004
DIS     0.0476  0.016
'''

In [19]:
df = pd.read_csv(io.StringIO(data_string), sep='\s+')
df

coef  std.error
Intercept  6.1529      0.072
CRIM      -0.0135      0.004
DIS        0.0476      0.016

In [20]:
df['t_nabluedaemoe'] = df['coef']/df['std.error']
df

coef  std.error  t_nabluedaemoe
Intercept  6.1529      0.072       85.456944
CRIM      -0.0135      0.004       -3.375000
DIS        0.0476      0.016        2.975000

In [23]:
t_critical = stats.t.ppf(1-0.05/2, 506 - (2+1))
t_critical

1.964691405362833

In [26]:
df['Znachimosty'] = df.apply(lambda x: 'znachim' if np.abs(x['t_nabluedaemoe']) >t_critical else 'ne znachim', axis = 1)
df

coef  std.error  t_nabluedaemoe Znachimosty
Intercept  6.1529      0.072       85.456944     znachim
CRIM      -0.0135      0.004       -3.375000     znachim
DIS        0.0476      0.016        2.975000     znachim

### Значимыми являются все коэффициенты регрессии